### ISSUE TITLE
Accessing s3 bucket on amazon cloud from lambda function fails with NoSuchBucket error

### ISSUE THREAD

**Issue by petr-pokorny-1** [localstack/localstack#11110]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When running the following code outside of localstack, it downloads file content from the s3 bucket on aws cloud. When it runs in lambda in localstack, the code fails with `error: NoSuchBucket: The specified bucket does not exist`
The configuration for s3Client explicitly defines `endpoint: 'https://s3.amazonaws.com'` so it should reach to real aws s3 service.

```typescript
        endpoint: 'https://s3.amazonaws.com',
        region: 'us-east-1',
        logger: console,
        credentials: {
            accessKeyId: "<access key id>",
            secretAccessKey: "<secret>"
        }
    }

    const getObjectParams = {
        Bucket: '<bucket name>',
        Key: '<key>'
    }

    const s3 = new S3Client(s3Config)
    const response = await s3.send(new GetObjectCommand(getObjectParams))
    const content = await response.Body?.transformToString()
```

### Expected Behavior

The file should be downloaded. Instead, the following error occurs:

`error: NoSuchBucket: The specified bucket does not exist`

### How are you starting LocalStack?

With a docker-compose file

#### How are you starting localstack (e.g., `bin/localstack` command, arguments, or `docker-compose.yml`)

    docker compose up with docker-compose.yml

#### Client commands (e.g., AWS SDK code snippet, or sequence of "awslocal" commands)

```yaml
services:
  localstack:
    container_name: "${LOCALSTACK_CONTAINER_NAME:-localstack}"
    image: "localstack/localstack-pro:${LOCALSTACK_VERSION:-latest}"  # required for Pro
    ports:
      - "127.0.0.1:4566:4566"            # LocalStack Gateway
      - "127.0.0.1:4510-4559:4510-4559"  # external services port range
      - "127.0.0.1:53:53"                # DNS config (required for Pro)
      - "127.0.0.1:443:443"              # LocalStack HTTPS Gateway (required for Pro)
    environment:
      - DEBUG=${DEBUG:-1}
      - PERSISTENCE=${PERSISTENCE:-0}
      - USE_SSL=${USE_SSL:-1}
      - LOCALSTACK_AUTH_TOKEN=${LOCALSTACK_AUTH_TOKEN-}
      - DOCKER_HOST=unix:///var/run/docker.sock
      - ECS_REMOVE_CONTAINERS=${ECS_REMOVE_CONTAINERS:-0}
      - DISABLE_EVENTS=${DISABLE_EVENTS:-0} # disable usage tracking
      - DYNAMODB_SHARE_DB=${DYNAMODB_SHARE_DB:-1}
      - DISABLE_CORS_CHECKS=1
      - DOCDB_PROXY_CONTAINER=1
      - LAMBDA_DISABLE_AWS_ENDPOINT_URL=1
    volumes:
      - "${LOCALSTACK_VOLUME_DIR:-./.localstack-volume}:/var/lib/localstack"
      - "/var/run/docker.sock:/var/run/docker.sock"
    healthcheck:
      test: ["CMD", "curl", "http://localhost:4566/_localstack/health?reload"]
      interval: 10s
      timeout: 5s
      retries: 10

networks:
  default:
    driver: bridge
```

### Environment

```markdown
- OS: MacOS Sonoma 14.5
- LocalStack:
LocalStack version: 3.5.1.dev20240624022330
LocalStack build date: 2024-06-24
LocalStack build git hash: 51dd910
```

### Anything else?

I also tried [MinIO client](https://github.com/minio/minio-js), but getting the same result - it downloads file from s3 when it runs outside of localstack, but fails on NoSuchBucket error when running in the lambda in localstack.

```typescript
  const minioClient = new Client({
    endPoint: 's3.amazonaws.com',
    useSSL: true,
    accessKey: '...',
    secretKey: '...'
  })

  await minioClient.fGetObject('bucket name', 'key', '<filePath>')
```

**Comment by cloutierMat** [comment]
Hi @petr-pokorny-1,

Thank you for your detailed report. Transparent endpoint injection is an expected behaviour in LocalStack, for more information you can read our documentation on [Transparent Endpoint Injection](https://docs.localstack.cloud/user-guide/tools/transparent-endpoint-injection/) and [Skip LocalStack DNS resolution](https://docs.localstack.cloud/user-guide/tools/dns-server/#skip-localstack-dns-resolution).

Let us know if we can help any further.

**Comment by petr-pokorny-1** [comment]
> Hi @petr-pokorny-1,
> 
> Thank you for your detailed report. Transparent endpoint injection is an expected behaviour in LocalStack, for more information you can read our documentation on [Transparent Endpoint Injection](https://docs.localstack.cloud/user-guide/tools/transparent-endpoint-injection/) and [Skip LocalStack DNS resolution](https://docs.localstack.cloud/user-guide/tools/dns-server/#skip-localstack-dns-resolution).
> 
> Let us know if we can help any further.

Thank you very much! This is super helpful. I have used `DNS_NAME_PATTERNS_TO_RESOLVE_UPSTREAM=my-bucket.s3.amazonaws.com` to bypass localstack endpoint injection for my bucket which I need to access on aws. Along with `forcePathStyle: false` in s3 client config. Problem solved.

**Comment by jonahelbaz-x** [comment]
> > Hi @petr-pokorny-1,
> > Thank you for your detailed report. Transparent endpoint injection is an expected behaviour in LocalStack, for more information you can read our documentation on [Transparent Endpoint Injection](https://docs.localstack.cloud/user-guide/tools/transparent-endpoint-injection/) and [Skip LocalStack DNS resolution](https://docs.localstack.cloud/user-guide/tools/dns-server/#skip-localstack-dns-resolution).
> > Let us know if we can help any further.
> 
> Thank you very much! This is super helpful. I have used `DNS_NAME_PATTERNS_TO_RESOLVE_UPSTREAM=my-bucket.s3.amazonaws.com` to bypass localstack endpoint injection for my bucket which I need to access on aws. Along with `forcePathStyle: false` in s3 client config. Problem solved.

I was hoping this comment was going to put an end to 8 hours of debugging

Maybe you can help? Within my localstack environment I have:

```
async getClient(): Promise<S3Client> {
    if (!this._client) {
      const region = process.env.AWS_S3_REGION;
      const accessKeyId = process.env.AWS_S3_ACCESS_KEY_ID;
      const secretAccessKey = process.env.AWS_S3_SECRET_ACCESS_KEY;

      this._client = new S3Client({
        // not valid in a lambda
        // useAccelerateEndpoint: true,
        credentials: {
          accessKeyId,
          secretAccessKey,
        },
        region,
        forcePathStyle: false,
      });
    }
    ```
    
    And within my function definition
    
    ```
        const environment = {
      POSTGRES_URL: Env.POSTGRES_URL,
      POSTGRES_SSL: Env.POSTGRES_SSL,
      AWS_S3_REGION: process.env.AWS_S3_REGION,
      AWS_S3_ACCESS_KEY_ID: process.env.AWS_S3_ACCESS_KEY_ID,
      AWS_S3_SECRET_ACCESS_KEY: process.env.AWS_S3_SECRET_ACCESS_KEY,
      DISABLE_TRANSPARENT_ENDPOINT_INJECTION: '1',
    };

    // create the avatar lambdas
    const inventoryForgeLambda = new NodejsFunction(
      scope,
      'TestLambda',
      {
        environment,
        ...
        ```
        
        
        Any help or guidance would be greatly appreciated!

**Comment by cloutierMat** [comment]
Hi @jonahelbaz-x,

Please not that `DNS_NAME_PATTERNS_TO_RESOLVE_UPSTREAM` is not a config used by the lambda executor, but rather configures LocalStack DNS behavior. It must be part of the LocalStack environment at launch (in docker-compose or env for the cli, etc).

I would advise against using a broad matching such as `*.s3.*.amazonaws.com` that could lead to any and all s3 calls to resolve in aws instead of LocalStack. Instead use narrower matching to reduce risk of wrongly making calls to AWS. For example `DNS_NAME_PATTERNS_TO_RESOLVE_UPSTREAM=my-bucket.s3.amazonaws.com`.

I hope that helps!

**Comment by JonahElbaz** [comment]
That was very helpful! Thank you!

**Comment by da-luce** [comment]
I have a similar setup where I am trying to access a remote DynamoDB table from a non-pro LocalStack lambda:

```python
    # Create DynamoDB client
    dynamodb = boto3.client(
        'dynamodb',
        region_name='us-east-1',
        endpoint_url='https://dynamodb.us-east-1.amazonaws.com'
    )
```

I have `DNS_NAME_PATTERNS_TO_RESOLVE_UPSTREAM='.*dynamodb.*.amazonaws.com'` and credentials properly configured via environment variables within the container. To be clear, this works as expected.

However, without explicitly setting the endpoint to `endpoint_url='https://dynamodb.us-east-1.amazonaws.com'`, boto3 resolves the endpoint to be `http://172.18.0.2:4566`. Why is this and how can I avoid explicitly setting the endpoint_url within my lambdas?
